### DISCUSSION TITLE
Option for Zoom Offset on Tile Layers

### DISCUSSION THREAD

**Discussion by program-sam** [openglobus/openglobus#634] [Ideas]
What would it take for tile layers to be able to use a zoom offset? What I mean by that is when a standard tile layer would be at zoom level 10, I would want a specific tile layer to be at zoom level 8 when the zoom offset is 2 for example.

I am currently looking at this for the CanvasTile layer, but this could be a feature for any tile based layer really.

I have been searching if I could make the suggested changes myself but could not seem to find how this is handled exactly. Is that something that would be done in the quadTree, changing when a node creates child nodes, and allowing that to be changed by a parameter? Or is the quadtree a global thing and is making it layer-specific not possible?

I'm curious to hear what you think of it.

**Comment by Zemledelec** [comment]
Hi @program-sam, could you provide a real example, I don't understand...:(

**Reply by program-sam** [reply]
Apologies if I wasn't able to make myself clear. This is a current view of what zoomlevel of tiles is loaded at a certain location at a certain distance above the ground (zoomLevel 8). I would like to be able to specify a zoomOffset, so that at this location, at this elevation above the ground, tiles of zoomlevel 8 - zoomOffset are shown. So in this example an offset so that tiles of zoomlevel 7 are shown instead of 8, a zoomOffset of 1.

![globus zoom](https://user-images.githubusercontent.com/75586344/220276474-7e9cce96-8875-4e38-a600-8b8ec152844d.png)

[(codepen)](https://codepen.io/program-sam/pen/ZEMQLBx)

The reason for this is that I am building custom canvas tiles using vector data, and it is computationally cheaper for me to generate fewer tiles at higher resolution, than to generate 4 times as many smaller tiles.

I thought it should be possible in theory because [leaflet had a similar feature](https://leafletjs.com/reference.html#tilelayer), but turns out that only offsets the zoom level that is used in the tile url, but doesn't correct the x & y or anything. So not sure if this is feasible with the way this is programmed

**Reply by Zemledelec** [reply]
Thank you! I've got it, yes leaflet and other tile engines have this feature. Let me think...

**Reply by Zemledelec** [reply]
There is a question
![image](https://user-images.githubusercontent.com/5452251/220324679-ad806627-e660-4192-a6ad-f7ccfa319c05.png)
What image zoom do we have to use for the smallest nodes?

**Reply by program-sam** [reply]
Still not sure if it is entirely clear, but I wouldn't expect different images on the tiles when using a tileOffset, I would expect the tile to just split at different elevations. With an offset of 1, it would split from level 1 to level 2 where the regular tree would split from 2 to 3. This way you end up with fewer, but bigger tiles in the map view.

![Untitled Diagram(1)](https://user-images.githubusercontent.com/75586344/220331271-2c2a7812-a54c-4cfa-9b4a-67f3646e000f.png)

**Comment by Zemledelec** [comment]
Hey @program-sam! Long time, sorry. There is no simple way to prevent quadtree traversing, but. As far as I understand, we are talking about the same thing. It's not hard to implement this feature. It includes two parts. First is to understand when to say XYZ layer applyMaterial method to stop load data, and second understand how to load data for the current node. For me first, and second parts are the same questions, the key answers lay inside applyMaterial and Material structure. I think this is doable. I'll check the code, and describe the overall algorithm, which could help to understand the idea. Might be the solution on the surface.

**Reply by program-sam** [reply]
Thanks for the response! No worries, ok good to know that we are on the same page. If you could describe an overall algorithm I might be able to suggest the code myself. But at this point I didn't understand the quad tree code thoroughly enough to make a pull request already. The quadtree traversal, is that something that happens for each layer individually, or is that a global thing and is there only one quad tree? That is something I couldn't figure out yet, but would be the difference between a layer-specific tileOffsets, or that a tileOffset would be a global parameter applying to all layers.

**Comment by program-sam** [comment]
i found that in the [Node component of the quadTree, at line 252](https://github.com/openglobus/openglobus/blob/master/src/og/quadTree/Node.js#L252) the size of the node is compared to the lodSize. This determines whether the node is rendered, or traversed further. By adding a zoomOffset parameter and manipulating this comparison, the desired result is achieved.

This allows for a global zoomOffset, not a layer-specific one, but this seemed an easy solution and might be enough for most / me

[Pull Request](https://github.com/openglobus/openglobus/pull/650)

**Reply by Zemledelec** [reply]
Hey @program-sam, could you answer in your PR thread, please?
